ast/src/analysis/mod.rs

```diff
@@ -0,0 +1,106 @@
+mod display;
+
+use std::collections::BTreeSet;
+
+use num_bigint::BigUint;
+
+use crate::parsed::{
+    asm::{InstructionBodyElement, InstructionParams, RegisterFlag},
+    Expression, Statement,
+};
+
+pub struct RegisterDeclarationStatement {
+    pub start: usize,
+    pub name: String,
+    pub flag: Option<RegisterFlag>,
+}
+
+pub struct InstructionDefinitionStatement<T> {
+    pub start: usize,
+    pub name: String,
+    pub params: InstructionParams,
+    pub body: Vec<InstructionBodyElement<T>>,
+}
+
+pub struct DegreeStatement {
+    pub degree: BigUint,
+}
+
+pub enum ProgramStatement<T> {
+    Assignment(AssignmentStatement<T>),
+    Instruction(InstructionStatement<T>),
+    Label(LabelStatement),
+}
+
+impl<T> From<AssignmentStatement<T>> for ProgramStatement<T> {
+    fn from(value: AssignmentStatement<T>) -> Self {
+        Self::Assignment(value)
+    }
+}
+
+impl<T> From<InstructionStatement<T>> for ProgramStatement<T> {
+    fn from(value: InstructionStatement<T>) -> Self {
+        Self::Instruction(value)
+    }
+}
+
+impl<T> From<LabelStatement> for ProgramStatement<T> {
+    fn from(value: LabelStatement) -> Self {
+        Self::Label(value)
+    }
+}
+
+pub struct AssignmentStatement<T> {
+    pub start: usize,
+    pub lhs: Vec<String>,
+    pub using_reg: Option<String>,
+    pub rhs: Box<Expression<T>>,
+}
+
+pub struct InstructionStatement<T> {
+    pub start: usize,
+    pub instruction: String,
+    pub inputs: Vec<Expression<T>>,
+}
+
+pub struct LabelStatement {
+    pub start: usize,
+    pub name: String,
+}
+
+pub struct PilBlock<T> {
+    pub start: usize,
+    pub statements: Vec<Statement<T>>,
+}
+
+pub struct AnalysisASMFile<T> {
+    pub degree: Option<DegreeStatement>,
+    pub registers: Vec<RegisterDeclarationStatement>,
+    pub pil: Vec<PilBlock<T>>,
+    pub instructions: Vec<InstructionDefinitionStatement<T>>,
+    pub program: Vec<ProgramStatement<T>>,
+    pub batches: Option<Vec<BatchMetadata>>,
+}
+
+#[derive(Default, Debug, PartialEq, Eq, Clone)]
+pub struct BatchMetadata {
+    // the set of compatible statements
+    pub size: usize,
+    // the reason why this batch ended (for debugging purposes), None if we ran out of statements to batch
+    pub reason: Option<IncompatibleSet>,
+}
+
+impl BatchMetadata {
+    pub fn get_size(&self) -> usize {
```

Comment:
```suggestion
    pub fn size(&self) -> usize {
```
personal taste, usually I don't like `get*` named functions